cv32e40s/tests/programs/custom/debug_priv_test/debug_priv_test.c

```diff
@@ -0,0 +1,260 @@
+// external assembly symbol which sets core into user mode.
+extern volatile void set_u_mode();
+// mstatus.MPP bit-field
+int MPP_FIELD [2] = {11, 12}; 
+// dcsr.prv bit-field
+int PRV_FIELD [2] = {0, 1}; 
+
+// declaration of the assert function
+static void assert_or_die(uint32_t actual, uint32_t expect, char *msg) {
+  if (actual != expect) {
+    printf(msg);
+    printf("expected = 0x%lx (%ld), got = 0x%lx (%ld)\n", expect, (int32_t)expect, actual, (int32_t)actual);
+    exit(EXIT_FAILURE);
+  }
+}
+
+/* 
+Retuns specific bit-field from [bit_indx_low : bit_indx_high] in register x
+*/
+unsigned int get_field(unsigned int x, int bit_indx_low, int bit_indx_high){
+    int field = ( 1 << ( (bit_indx_high - bit_indx_low) + 1) )  - 1;
+    x = (x & (field << bit_indx_low) ) >> bit_indx_low;
+    return x;
+}
+
+
+/*
+Checks the mepc for compressed instructions and increments appropriately 
+*/
+void increment_mepc(void){
+  volatile unsigned int insn, mepc;
+
+    __asm__ volatile("csrrs %0, mepc, x0" : "=r"(mepc)); // read the mepc
+    __asm__ volatile("lw %0, 0(%1)" : "=r"(insn) : "r"(mepc)); // read the contents of the mepc pc.
+    if ((insn & 0x3) == 0x3) { // check for compressed instruction before increment.
+      mepc += 4;
+    } else {
+      mepc += 2;
+    }
+    __asm__ volatile("csrrw x0, mepc, %0" :: "r"(mepc)); // write to the mepc 
+}
+
+
+// Rewritten interrupt handler
+__attribute__ ((interrupt ("machine")))
+void u_sw_irq_handler(void) {
+
+    __asm__ volatile("csrrs %0, mcause, x0" : "=r"(wmcause));
+    __asm__ volatile("csrrs %0, mstatus, x0" : "=r"(wmstatus));
+    __asm__ volatile("csrrw x0, mstatus, %0" :: "r"(MSTATUS_STD_VAL)); // set machine mode 
+    increment_mepc();
+}
+
+// Declaration of the debug struct 
+typedef union {
+  struct {
+    unsigned int start_delay      : 15; // 14: 0
+    unsigned int rand_start_delay : 1;  //    15
+    unsigned int pulse_width      : 13; // 28:16
+    unsigned int rand_pulse_width : 1;  //    29
+    unsigned int pulse_mode       : 1;  //    30    0 = level, 1 = pulse
+    unsigned int value            : 1;  //    31
+  } fields;
+  unsigned int bits;
+}  debug_req_control_t;
+
+/*
+When called it starts debug_mode
+*/
+void run_debug_mode(void) {
+  // debug control struct 
+  debug_req_control_t debug_req_control;
+  debug_req_control = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1, //PULSE Mode
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 5,// FIXME: BUG: one clock pulse cause core to lock up
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  // this will initiate debug mode 
+  DEBUG_REQ_CONTROL_REG = debug_req_control.bits; 
+  // waits for debug to finish
+  while(glb_debug_status != 1){
+      ;
+  }
+  // resets the debug status bit
+  glb_debug_status = 0;
+}
+
+int main(void){// TODO: test will failed until issue #278 in core-v-verif/cv32e40s is fixed
```

Comment:
Will the test fail when run, or are the expected-to-fail parts just commented out?
(I'd prefer the latter, with a todo at the place where code is commented out.)

Reply:
The actual tests will pass, but it will print FAILED at the end because of the assert issue in #278

Reply:
Right. That is alright then.